File: typescript/src/gonkaOpenAI.ts
import OpenAI from 'openai';
import { GonkaOpenAIOptions, SignatureComponents, GonkaEndpoint } from './types.js';
import { customEndpointSelection, gonkaBaseURL, gonkaFetch, getNanoTimestamp, resolveEndpoints } from './utils.js';
import { gonkaSignature as signatureFunction } from './utils.js';
import { ENV } from './constants.js';

/**
 * GonkaOpenAI client that extends the official OpenAI client to work with Gonka network
 */
export class GonkaOpenAI extends OpenAI {
  private readonly _privateKey: string;
  
  // Store the selected endpoint
  private readonly _selectedEndpoint: GonkaEndpoint;
  private _sourceUrl?: string;
  private _endpoints?: GonkaEndpoint[];
  
  /**
   * Create a new GonkaOpenAI client
   * 
   * @param options Options for the client
   */
  constructor(options: GonkaOpenAIOptions) {
    // Get private key from options or environment
    const privateKey = options.gonkaPrivateKey || process.env[ENV.PRIVATE_KEY];
    if (!privateKey) {
      throw new Error(`Private key must be provided either in options or through ${ENV.PRIVATE_KEY} environment variable`);
    }

    // Prepare configuration (avoid using `this` before super())
    const sourceUrlLocal = options.sourceUrl || process.env['GONKA_SOURCE_URL'];
    const endpointsLocal: GonkaEndpoint[] | undefined =
      options.endpoints && options.endpoints.length ? options.endpoints : undefined;

    // Determine candidate endpoints for selection (only use provided; env/defaults handled by gonkaBaseURL)
    const resolvedEndpoints: GonkaEndpoint[] = endpointsLocal || [];

    // Determine the endpoint
    let selectedEndpoint: GonkaEndpoint;
    if (options.endpointSelectionStrategy) {
      // Use custom endpoint selection strategy if provided
      selectedEndpoint = customEndpointSelection(options.endpointSelectionStrategy, resolvedEndpoints);
    } else {
      // Default to random endpoint selection
      selectedEndpoint = gonkaBaseURL(resolvedEndpoints);
    }

    // Create the signing fetch function
    const signingFetch = gonkaFetch({
      gonkaPrivateKey: privateKey,
      gonkaAddress: options.gonkaAddress || process.env[ENV.ADDRESS],
      selectedEndpoint: selectedEndpoint,
    });

    // Normalize endpoint alias for consistency
    if (!selectedEndpoint.address) selectedEndpoint.address = selectedEndpoint.transferAddress;

    // Create the OpenAI configuration object
    const openAIConfig = {
      ...options,
      baseURL: selectedEndpoint.url,
    };
    
    // Set default mock-api-key if no apiKey is provided
    if (!openAIConfig.apiKey) {
      openAIConfig.apiKey = "mock-api-key";
    }
    
    // Add the signing fetch function to the configuration
    (openAIConfig as any).fetch = signingFetch;

    // Call OpenAI constructor with the configuration
    super(openAIConfig);
    
    // Save the private key and selected endpoint for signing requests
    this._privateKey = privateKey;
    this._selectedEndpoint = selectedEndpoint;
    this._sourceUrl = sourceUrlLocal;
    this._endpoints = endpointsLocal;
  }
  
  /**
   * Get the private key
   */
  get privateKey(): string {
    return this._privateKey;
  }

  /**
   * Sign a request body with the client's private key
   * 
   * @param body The request body to sign
   * @param transferAddress The Cosmos address of the endpoint provider (optional, uses the selected endpoint's address if not provided)
   * @returns The signature as a base64 string
   */
  async signRequest(body: any, transferAddress?: string): Promise<string> {
    // Generate a unique timestamp in nanoseconds
    const timestamp = getNanoTimestamp();
    
    // Use the provided transfer address or the selected endpoint's address
    const address = transferAddress || this._selectedEndpoint.address || this._selectedEndpoint.transferAddress;
    
    // Create signature components
    const components: SignatureComponents = {
      payload: body,
      timestamp: timestamp,
      transferAddress: address
    };
    
    return await signatureFunction(components, this._privateKey);
  }
  
  /**
   * Get the selected endpoint
   */
  get selectedEndpoint() {
    return this._selectedEndpoint;
  }

  /**
   * Async factory method that resolves endpoints with filtering by
   * allowed_transfer_addresses and delegate_ta preference before constructing the client.
   */
  static async create(options: GonkaOpenAIOptions): Promise<GonkaOpenAI> {
    const resolved = await resolveEndpoints({
      sourceUrl: options.sourceUrl,
      endpoints: options.endpoints,
    });
    return new GonkaOpenAI({
      ...options,
      endpoints: resolved,
      // Clear sourceUrl so the constructor doesn't try to re-resolve
      sourceUrl: undefined,
    });
  }
}

File: go/gonkaopenai.go
package gonkaopenai

import (
	"context"
	"fmt"
	"net/http"
	"os"
	"strings"

	"github.com/openai/openai-go"
	"github.com/openai/openai-go/option"
)

// Endpoint represents a Gonka API endpoint with its associated transfer address
type Endpoint struct {
	URL     string
	Address string
	Weight  int64
}

// Options for creating a GonkaOpenAI client.
type Options struct {
	APIKey                    string
	GonkaPrivateKey           string
	GonkaAddress              string
	EndpointSelectionStrategy func([]Endpoint) string
	HTTPClient                *http.Client
	OrgID                     string
	SourceUrl                 string
	Endpoints                 []Endpoint
}

// GonkaOpenAI wraps the official openai.Client.
type GonkaOpenAI struct {
	*openai.Client
	privateKey string
	gonkaAddr  string
}

// NewGonkaOpenAI creates a new client configured for the Gonka network.
func NewGonkaOpenAI(opts Options) (*GonkaOpenAI, error) {
	privateKey := opts.GonkaPrivateKey
	if privateKey == "" {
		privateKey = os.Getenv(EnvPrivateKey)
	}
	if privateKey == "" {
		return nil, fmt.Errorf("private key must be provided via opts or %s", EnvPrivateKey)
	}

	// Determine endpoints per priority:
	// 1) SourceUrl in opts or env -> try fetch participants
	// 2) If no endpoints from SourceUrl, use opts.Endpoints
	// 3) If still none, try env GONKA_ENDPOINTS (via utils not available here; expect caller to provide)
	sourceUrl := opts.SourceUrl
	if sourceUrl == "" {
		sourceUrl = os.Getenv(EnvSourceUrl)
	}

	var endpoints []Endpoint
	if sourceUrl != "" {
		eps, err := GetParticipantsWithProof(context.Background(), sourceUrl, "current")
		if err == nil && len(eps) > 0 {
			endpoints = eps
		}
	}
	if len(endpoints) == 0 && len(opts.Endpoints) > 0 {
		endpoints = opts.Endpoints
	}
	if len(endpoints) == 0 {
		endpoints = GetEndpointsFromEnv()
	}
	if len(endpoints) == 0 {
		return nil, fmt.Errorf("no endpoints resolved from SourceUrl, Options.Endpoints, or %s", EnvEndpoints)
	}

	// Validate that each endpoint has a non-empty address
	for _, endpoint := range endpoints {
		if endpoint.Address == "" {
			return nil, fmt.Errorf("endpoint %s has an empty address, all endpoints must have an address", endpoint.URL)
		}
	}

	baseURL := ""
	if opts.EndpointSelectionStrategy != nil {
		baseURL = CustomEndpointSelection(opts.EndpointSelectionStrategy, endpoints)
	} else {
		baseURL = GonkaBaseURL(endpoints)
	}

	address := opts.GonkaAddress
	if address == "" {
		address = os.Getenv(EnvAddress)
	}
	if address == "" {
		addr, err := GonkaAddress(privateKey)
		if err == nil {
			address = addr
		} else {
			prefix := strings.Split(GonkaChainID, "-")[0]
			if len(privateKey) > 40 {
				address = fmt.Sprintf("%s1%s", prefix, privateKey[:40])
			} else {
				address = fmt.Sprintf("%s1%s", prefix, privateKey)
			}
		}
	}

	// Create HTTP client with endpoints
	httpClient, err := GonkaHTTPClient(HTTPClientOptions{
		PrivateKey: privateKey,
		Address:    address,
		Endpoints:  endpoints,
		Client:     opts.HTTPClient,
	})
	if err != nil {
		return nil, fmt.Errorf("failed to create HTTP client: %w", err)
	}

	var clientOptions []option.RequestOption
	clientOptions = append(clientOptions, option.WithBaseURL(baseURL))
	clientOptions = append(clientOptions, option.WithHTTPClient(httpClient))

	apiKey := opts.APIKey
	if apiKey == "" {
		apiKey = "mock-api-key" // Default mock API key if not provided
	}
	clientOptions = append(clientOptions, option.WithAPIKey(apiKey))

	if opts.OrgID != "" {
		clientOptions = append(clientOptions, option.WithOrganization(opts.OrgID))
	}

	rawClient := openai.NewClient(clientOptions...)
	return &GonkaOpenAI{Client: &rawClient, privateKey: privateKey, gonkaAddr: address}, nil
}

// GonkaAddress returns the configured Gonka address.
func (g *GonkaOpenAI) GonkaAddress() string { return g.gonkaAddr }

// PrivateKey returns the private key used for signing.
func (g *GonkaOpenAI) PrivateKey() string { return g.privateKey }

// ExampleChatCompletion demonstrates a simple call using the Gonka client.
func ExampleChatCompletion(ctx context.Context, g *GonkaOpenAI) (*openai.ChatCompletion, error) {
	return g.Chat.Completions.New(ctx, openai.ChatCompletionNewParams{
		Model: "Qwen/QwQ-32B",
		Messages: []openai.ChatCompletionMessageParamUnion{
			openai.UserMessage("Hello!"), // Use UserMessage helper
		},
	})
}
